File: core/plugin/__DEV/Python/n_gfRigPSDVectorAngle.py
# -*- coding: utf-8 -*-
"""
Copyright (c) 2019 Giuliano França

Redistribution:
    Something here.

Maya Node:
    [This is a prototype version of the gfRigPSDVectorAngle node. You should be using the related C++ version.]
    This node is a test node and only performs test operations with one input value.
    https://www.desmos.com/calculator/nfggjvzpkn

Requirements:
    Maya 2017 or above.

Todo:
    * NDA

This code supports Pylint. Rc file in project.
"""
# pylint: disable=E0401
# E0401 = Supress Maya modules import error

import math
import maya.api.OpenMaya as om2


def maya_useNewAPI():
    """ Function to Maya recognize the use of the Python API 2.0. """
    # pylint: disable=C0103, w0107
    pass


def INPUT_ATTR(FNATTR):
    """ Configure a input attribute. """
    # pylint: disable=C0103
    FNATTR.writable = True
    FNATTR.readable = True
    FNATTR.storable = True
    FNATTR.keyable = True


def OUTPUT_ATTR(FNATTR):
    """ Configure a output attribute. """
    # pylint: disable=C0103
    FNATTR.writable = False
    FNATTR.readable = True
    FNATTR.storable = False
    FNATTR.keyable = False


class VectorAnglePSD(om2.MPxNode):
    """ Main class of gfRigPSDVectorAngle node. """

    kNODE_NAME = ""
    kNODE_CLASSIFY = ""
    kNODE_ID = ""

    inBase = om2.MObject()
    inSource = om2.MObject()
    inTarget = om2.MObject()
    inTargetEnvelope = om2.MObject()
    inTargetFalloff = om2.MObject()
    inRampWeights = om2.MObject()
    outWeights = om2.MObject()

    def __init__(self):
        """ Constructor. """
        om2.MPxNode.__init__(self)

    def postConstructor(self):
        """ Post Constructor. """
        thisMob = self.thisMObject()
        attr = VectorAnglePSD.inRampWeights
        ramp = om2.MRampAttribute(thisMob, attr)
        pos = om2.MFloatArray()
        val = om2.MFloatArray()
        interp = om2.MIntArray()
        pos.append(1.0)
        val.append(1.0)
        interp.append(om2.MRampAttribute.kLinear)
        ramp.addEntries(pos, val, interp)

    @staticmethod
    def creator():
        """ Maya creator function. """
        return VectorAnglePSD()

    @staticmethod
    def initialize():
        """
        Defines the set of attributes for this node. The attributes declared in this function are assigned
        as static members to IKVChain class. Instances of IKVChain will use these attributes to create plugs
        for use in the compute() method.
        """
        mAttr = om2.MFnMatrixAttribute()
        nAttr = om2.MFnNumericAttribute()
        rAttr = om2.MRampAttribute()

        VectorAnglePSD.inBase = mAttr.create("base", "base", om2.MFnMatrixAttribute.kFloat)
        INPUT_ATTR(mAttr)

        VectorAnglePSD.inSource = mAttr.create("source", "source", om2.MFnMatrixAttribute.kFloat)
        INPUT_ATTR(mAttr)

        VectorAnglePSD.inTarget = mAttr.create("target", "target", om2.MFnMatrixAttribute.kFloat)
        mAttr.array = True
        INPUT_ATTR(mAttr)

        VectorAnglePSD.inTargetEnvelope = nAttr.create("targetEnvelope", "te", om2.MFnNumericData.kFloat, 1.0)
        nAttr.setMin(0.0)
        nAttr.setMax(1.0)
        nAttr.array = True
        INPUT_ATTR(nAttr)

        VectorAnglePSD.inTargetFalloff = nAttr.create("targetFalloff", "tf", om2.MFnNumericData.kFloat, 90.0)
        nAttr.setMin(0.0)
        nAttr.setMax(180.0)
        nAttr.array = True
        INPUT_ATTR(nAttr)

        VectorAnglePSD.inRampWeights = rAttr.createCurveRamp("rampWeights", "rw")

        VectorAnglePSD.outWeights = nAttr.create("outWeights", "ow", om2.MFnNumericData.kFloat, 0.0)
        nAttr.array = True
        OUTPUT_ATTR(nAttr)

        VectorAnglePSD.addAttribute(VectorAnglePSD.inBase)
        VectorAnglePSD.addAttribute(VectorAnglePSD.inSource)
        VectorAnglePSD.addAttribute(VectorAnglePSD.inTarget)
        VectorAnglePSD.addAttribute(VectorAnglePSD.inTargetEnvelope)
        VectorAnglePSD.addAttribute(VectorAnglePSD.inTargetFalloff)
        VectorAnglePSD.addAttribute(VectorAnglePSD.inRampWeights)
        VectorAnglePSD.addAttribute(VectorAnglePSD.outWeights)
        VectorAnglePSD.attributeAffects(VectorAnglePSD.inBase, VectorAnglePSD.outWeights)
        VectorAnglePSD.attributeAffects(VectorAnglePSD.inSource, VectorAnglePSD.outWeights)
        VectorAnglePSD.attributeAffects(VectorAnglePSD.inTarget, VectorAnglePSD.outWeights)
        VectorAnglePSD.attributeAffects(VectorAnglePSD.inTargetEnvelope, VectorAnglePSD.outWeights)
        VectorAnglePSD.attributeAffects(VectorAnglePSD.inTargetFalloff, VectorAnglePSD.outWeights)
        VectorAnglePSD.attributeAffects(VectorAnglePSD.inRampWeights, VectorAnglePSD.outWeights)

    def compute(self, plug, dataBlock):
        """
        Node computation method:
            * plug is a connection point related to one of our node attributes (either an input or an output).
            * dataBlock contains the data on which we will base our computations.
        """
        # pylint: disable=R0201
        if plug != VectorAnglePSD.outWeights:
            return om2.kUnknownParameter

        mBase = dataBlock.inputValue(VectorAnglePSD.inBase).asFloatMatrix()
        mSource = dataBlock.inputValue(VectorAnglePSD.inSource).asFloatMatrix()

        targetHandle = dataBlock.inputArrayValue(VectorAnglePSD.inTarget)
        targetEnvelopeHandle = dataBlock.inputArrayValue(VectorAnglePSD.inTargetEnvelope)
        targetFalloffHandle = dataBlock.inputArrayValue(VectorAnglePSD.inTargetFalloff)
        outWeightsHandle = dataBlock.outputArrayValue(VectorAnglePSD.outWeights)

        vBase = om2.MFloatVector(mBase[12], mBase[13], mBase[14])
        vSource = om2.MFloatVector(mSource[12], mSource[13], mSource[14])

        vCurPose = vSource - vBase
        nCurPose = vCurPose.normal()

        targetList = []
        envelopeList = []
        falloffList = []

        for i in range(len(targetHandle)):
            targetHandle.jumpToLogicalElement(i)
            mtx = targetHandle.inputValue().asFloatMatrix()
            vec = om2.MFloatVector(mtx[12], mtx[13], mtx[14])
            vTargetPose = vec - vBase
            nTargetPose = vTargetPose.normal()
            targetList.append(nTargetPose)

        for i in range(len(targetEnvelopeHandle)):
            targetEnvelopeHandle.jumpToLogicalElement(i)
            env = targetEnvelopeHandle.inputValue().asFloat()
            envelopeList.append(env)

        for i in range(len(targetFalloffHandle)):
            targetFalloffHandle.jumpToLogicalElement(i)
            fall = targetFalloffHandle.inputValue().asFloat()
            falloffList.append(fall)

        for i in range(len(outWeightsHandle)):
            outWeightsHandle.jumpToLogicalElement(i)
            resultHandle = outWeightsHandle.outputValue()
            if (i < len(outWeightsHandle) and
                    i < len(targetHandle) and
                    i < len(targetEnvelopeHandle) and
                    i < len(targetFalloffHandle)):
                theta = math.acos(targetList[i] * nCurPose)
                ratio = min(max(theta/math.radians(falloffList[i]), -1.0), 1.0)
                if ratio == 0.0:
                    weight = envelopeList[i] * 1.0
                elif ratio > 0.0:
                    weight = envelopeList[i] * (1.0 - ratio)
                elif ratio < 0.0:
                    weight = envelopeList[i] * (1.0 + ratio)

                thisMob = self.thisMObject()
                rampAttr = om2.MRampAttribute(thisMob, VectorAnglePSD.inRampWeights)
                resultWeight = rampAttr.getValueAtPosition(weight)
                resultHandle.setFloat(resultWeight)
            else:
                resultHandle.setFloat(0.0)

File: plug-ins/dev/cpp/source/n_gfRigPSDVectorAngle.cpp
#include "headers/n_gfRigPSDVectorAngle.h"

// Configure a input attribute.
#define INPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(true);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(true);	\
	FNATTR.setKeyable(true);	\

// Configure a output attribute.
#define OUTPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(false);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(false);	\
	FNATTR.setKeyable(false);	\


// Constructor.
VectorAnglePSD::VectorAnglePSD() {}

// Destructor.
VectorAnglePSD::~VectorAnglePSD() {}

MObject VectorAnglePSD::inBase;
MObject VectorAnglePSD::inSource;
MObject VectorAnglePSD::inTarget;
MObject VectorAnglePSD::inTargetEnvelope;
MObject VectorAnglePSD::inTargetFalloff;
MObject VectorAnglePSD::inRampWeights;
MObject VectorAnglePSD::outWeights;


static float degToRad(float deg){
    // Convert degrees to radians.
    return deg * (static_cast<float>(M_PI) / 180.0f);
}

void VectorAnglePSD::postConstructor(){
    // Post Constructor.
    MObject thisMob = thisMObject();
    MRampAttribute ramp = MRampAttribute(thisMob, inRampWeights);
    MFloatArray pos = MFloatArray();
    MFloatArray val = MFloatArray();
    MIntArray interp = MIntArray();
    pos.append(1.0f);
    val.append(1.0f);
    interp.append(MRampAttribute::kLinear);
    ramp.addEntries(pos, val, interp);
}

void* VectorAnglePSD::creator(){
    // Maya creator function.
    return new VectorAnglePSD();
}

MStatus VectorAnglePSD::initialize(){
    /*
    Defines the set of attributes for this node. The attributes declared in this function are assigned
    as static members to IKVChain class. Instances of IKVChain will use these attributes to create plugs
    for use in the compute() method.
    */
    MStatus status;
    MFnMatrixAttribute mAttr;
    MFnNumericAttribute nAttr;
    MRampAttribute rAttr;

    inBase = mAttr.create("base", "base", MFnMatrixAttribute::kFloat, &status);
    INPUT_ATTR(mAttr);

    inSource = mAttr.create("source", "source", MFnMatrixAttribute::kFloat, &status);
    INPUT_ATTR(mAttr);

    inTarget = mAttr.create("target", "target", MFnMatrixAttribute::kFloat, &status);
    mAttr.setArray(true);
    INPUT_ATTR(mAttr);

    inTargetEnvelope = nAttr.create("targetEnvelope", "te", MFnNumericData::kFloat, 1.0f, &status);
    nAttr.setMin(0.0f);
    nAttr.setMax(1.0f);
    nAttr.setArray(true);
    INPUT_ATTR(nAttr);

    inTargetFalloff = nAttr.create("targetFalloff", "tf", MFnNumericData::kFloat, 90.0f, &status);
    nAttr.setMin(0.0f);
    nAttr.setMax(180.0f);
    nAttr.setArray(true);
    INPUT_ATTR(nAttr);

    inRampWeights = rAttr.createCurveRamp("rampWeights", "rw", &status);

    outWeights = nAttr.create("outWeights", "ow", MFnNumericData::kFloat, 0.0f, &status);
    nAttr.setArray(true);
    OUTPUT_ATTR(nAttr);

    addAttribute(inBase);
    addAttribute(inSource);
    addAttribute(inTarget);
    addAttribute(inTargetEnvelope);
    addAttribute(inTargetFalloff);
    addAttribute(inRampWeights);
    addAttribute(outWeights);
    attributeAffects(inBase, outWeights);
    attributeAffects(inSource, outWeights);
    attributeAffects(inTarget, outWeights);
    attributeAffects(inTargetEnvelope, outWeights);
    attributeAffects(inTargetFalloff, outWeights);
    attributeAffects(inRampWeights, outWeights);

    return status;
}

MStatus VectorAnglePSD::compute(const MPlug& plug, MDataBlock& dataBlock){
    /*
    Node computation method:
        * plug is a connection point related to one of our node attributes (either an input or an output).
        * dataBlock contains the data on which we will base our computations.
    */
    if (plug != outWeights)
        return MStatus::kUnknownParameter;

    MFloatMatrix mBase = dataBlock.inputValue(inBase).asFloatMatrix();
    MFloatMatrix mSource = dataBlock.inputValue(inSource).asFloatMatrix();

    MArrayDataHandle targetHandle = dataBlock.inputArrayValue(inTarget);
    MArrayDataHandle targetEnvelopeHandle = dataBlock.inputArrayValue(inTargetEnvelope);
    MArrayDataHandle targetFalloffHandle = dataBlock.inputArrayValue(inTargetFalloff);
    MArrayDataHandle outWeightsHandle = dataBlock.inputArrayValue(outWeights);

    MFloatVector vBase = MFloatVector(mBase[3][0], mBase[3][1], mBase[3][2]);
    MFloatVector vSource = MFloatVector(mSource[3][0], mSource[3][1], mSource[3][2]);

    MFloatVector vCurPose = vSource - vBase;
    MFloatVector nCurPose = vCurPose.normal();

    std::vector<MFloatVector> targetList;
    std::vector<float> envelopeList;
    std::vector<float> falloffList;
    
    for (uint32_t i = 0; i < targetHandle.elementCount(); i++){
        targetHandle.jumpToArrayElement(i);
        MFloatMatrix mtx = targetHandle.inputValue().asFloatMatrix();
        MFloatVector vec = MFloatVector(mtx[3][0], mtx[3][1], mtx[3][2]);
        MFloatVector vTargetPose = vec - vBase;
        MFloatVector nTargetPose = vTargetPose.normal();
        targetList.push_back(nTargetPose);
    }

    for (uint32_t i = 0; i < targetEnvelopeHandle.elementCount(); i++){
        targetEnvelopeHandle.jumpToArrayElement(i);
        float env = targetEnvelopeHandle.inputValue().asFloat();
        envelopeList.push_back(env);
    }

    for (uint32_t i = 0; i < targetFalloffHandle.elementCount(); i++){
        targetFalloffHandle.jumpToArrayElement(i);
        float fall = targetFalloffHandle.inputValue().asFloat();
        falloffList.push_back(fall);
    }

    for(uint32_t i = 0; i < outWeightsHandle.elementCount(); i++){
        outWeightsHandle.jumpToArrayElement(i);
        MDataHandle resultHandle = outWeightsHandle.outputValue();
        if ((i < outWeightsHandle.elementCount()) &&
            (i < targetHandle.elementCount()) &&
            (i < targetEnvelopeHandle.elementCount()) &&
            (i < targetFalloffHandle.elementCount())){
                float theta = std::acosf(targetList[i] * nCurPose);
                float ratio = std::min(std::max(theta / degToRad(falloffList[i]), -1.0f), 1.0f);
                float weight;
                if (ratio == 0.0f)
                    weight = envelopeList[i];
                else if (ratio > 0.0f)
                    weight = envelopeList[i] * (1.0f - ratio);
                else if (ratio < 0.0f)
                    weight = envelopeList[i] * (1.0f + ratio);

                MObject thisMob = thisMObject();
                MRampAttribute rampAttr = MRampAttribute(thisMob, inRampWeights);
                float resultWeight;
                rampAttr.getValueAtPosition(weight, resultWeight);
                resultHandle.setFloat(resultWeight);
            }
        else
            resultHandle.setFloat(0.0f);
        
        resultHandle.setClean();
    }

    return MStatus::kSuccess;
}